poly/src/domain/radix2/mod.rs

```diff
@@ -127,87 +149,16 @@ impl<F: FftField> EvaluationDomain<F> for Radix2EvaluationDomain<F> {
-        //
-        // However, if tau in H, both the numerator and denominator equal 0
-        // when i corresponds to the value tau equals, and the coefficient is 0
-        // everywhere else. We handle this case separately, and we can easily
-        // detect by checking if the vanishing poly is 0.
-        let size = self.size();
-        // TODO: Make this use the vanishing polynomial
-        let z_h_at_tau = tau.pow(&[self.size]) - F::one();
-        let domain_offset = F::one();
-        if z_h_at_tau.is_zero() {
-            // In this case, we know that tau = hg^i, for some value i.
-            // Then i-th lagrange coefficient in this case is then simply 1,
-            // and all other lagrange coefficients are 0.
-            // Thus we find i by brute force.
-            let mut u = vec![F::zero(); size];
-            let mut omega_i = domain_offset;
-            for u_i in u.iter_mut().take(size) {
-                if omega_i == tau {
-                    *u_i = F::one();
-                    break;
-                }
-                omega_i *= &self.group_gen;
-            }
-            u
-        } else {
-            // In this case we have to compute `Z_H(tau) * v_i / (tau - h g^i)`
-            // for i in 0..size
-            // We actually compute this by computing (Z_H(tau) * v_i)^{-1} * (tau - h g^i)
-            // and then batch inverting to get the correct lagrange coefficients.
-            // We let `l_i = (Z_H(tau) * v_i)^-1` and `r_i = tau - h g^i`
-            // Notice that since Z_H(tau) is i-independent,
-            // and v_i = g * v_{i-1}, it follows that
-            // l_i = g^-1 * l_{i-1}
-            // TODO: consider caching the computation of l_i to save N multiplications
-            use ark_ff::fields::batch_inversion;
-
-            // v_0_inv = m * h^(m-1)
-            let v_0_inv = F::from(self.size) * domain_offset.pow(&[self.size - 1]);
-            let mut l_i = z_h_at_tau.inverse().unwrap() * v_0_inv;
-            let mut negative_cur_elem = -domain_offset;
-            let mut lagrange_coefficients_inverse = vec![F::zero(); size];
-            for coeff in &mut lagrange_coefficients_inverse {
-                let r_i = tau + negative_cur_elem;
-                *coeff = l_i * r_i;
-                // Increment l_i and negative_cur_elem
-                l_i *= &self.group_gen_inv;
-                negative_cur_elem *= &self.group_gen;
-            }
-
-            // Invert the lagrange coefficients inverse, to get the actual coefficients,
-            // and return these
-            batch_inversion(lagrange_coefficients_inverse.as_mut_slice());
-            lagrange_coefficients_inverse
-        }
-    }
-
     fn vanishing_polynomial(&self) -> crate::univariate::SparsePolynomial<F> {
-        let coeffs = vec![(0, -F::one()), (self.size(), F::one())];
+        let coeffs = vec![(0, -self.offset.pow([self.size])), (self.size(), F::one())];
```

Comment:
We should specialize this in the case that `self.offset.is_one()`, to avoid unnecessary work.
```suggestion
        let constant_coeff = if self.offset.is_one() { self.offset } else { self.offset.pow([self.size]) };
        let coeffs = vec![(0, -constant_coeff), (self.size(), F::one())];
```

Reply:
Another way around this would be to just store `self.offset.pow([self.size])` in the domain when constructing it (though this could hinder a default impl in the trait)

Reply:
Yeah, I prefer storing it in the domain